gssapi/WrapToken.go

```diff
@@ -0,0 +1,230 @@
+   in this document), which consists of a descriptive header, followed
+   by a body portion that contains either the input user data in
+   plaintext concatenated with the checksum, or the input user data
+   encrypted.  The GSS_Wrap() token SHALL have the following format:
+
+         Octet no   Name        Description
+         --------------------------------------------------------------
+          0..1     TOK_ID    Identification field.  Tokens emitted by
+                             GSS_Wrap() contain the hex value 05 04
+                             expressed in big-endian order in this
+                             field.
+          2        Flags     Attributes field, as described in section
+                             4.2.2.
+          3        Filler    Contains the hex value FF.
+          4..5     EC        Contains the "extra count" field, in big-
+                             endian order as described in section 4.2.3.
+          6..7     RRC       Contains the "right rotation count" in big-
+                             endian order, as described in section
+                             4.2.5.
+          8..15    SND_SEQ   Sequence number field in clear text,
+                             expressed in big-endian order.
+          16..last Data      Encrypted data for Wrap tokens with
+                             confidentiality, or plaintext data followed
+                             by the checksum for Wrap tokens without
+                             confidentiality, as described in section
+                             4.2.4.
+
+Quick notes:
+	- "EC" or "Extra Count" refers to the length of the cheksum.
+	- "Flags" (complete details in section 4.2.2) is a set of bits:
+		- if bit 0 is set, it means the token was sent by the acceptor (generally the kerberized service).
+		- bit 1 indicates that the token's payload is encrypted
+ 		- bit 2 indicates if the message is protected using a subkey defined by the acceptor.
+	- When computing checksums, EC and RRC MUST be set to 0.
+    - Wrap Tokens are not ASN.1 encoded.
+*/
+var (
+	HdrLen              = 16 // Length of the Wrap Token's header
+	GSSWrapTokenID      = [2]byte{0x05, 0x04}
+	FillerByte     byte = 0xFF
+	ChecksumECRRC       = [4]byte{0x00, 0x00, 0x00, 0x00}
+	ENC                 = binary.BigEndian
+)
+
+type WrapToken struct {
+	// const GSS Token ID: 0x0504
+	Flags byte // acceptor, sealed, acceptor subkey
+	// const Filler: 0xFF
+	EC       uint16 // checksum length. big-endian
+	RRC      uint16 // right rotation count. big-endian
+	SND_SEQ  uint64 // sender's sequence number. big-endian
+	Payload  []byte // your data! :)
+	CheckSum []byte // authenticated checksum of { payload | header }
+}
+
+// Get them bytes!
+func (wt *WrapToken) Marshal() ([]byte, error) {
+	if wt.CheckSum == nil {
+		return nil, errors.New("Checksum has not been set.")
+	}
+	if wt.Payload == nil {
+		return nil, errors.New("Payload has not been set.")
```

Comment:
best practice for Go is for error messages to start with a lower case character.